buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -229,54 +229,82 @@ public static int compare(ByteBuf bufferA, ByteBuf bufferB) {
         final int minLength = Math.min(aLen, bLen);
         final int uintCount = minLength >>> 2;
         final int byteCount = minLength & 3;
-
         int aIndex = bufferA.readerIndex();
         int bIndex = bufferB.readerIndex();
 
-        if (bufferA.order() == bufferB.order()) {
-            for (int i = uintCount; i > 0; i --) {
-                long va = bufferA.getUnsignedInt(aIndex);
-                long vb = bufferB.getUnsignedInt(bIndex);
-                if (va > vb) {
-                    return 1;
-                }
-                if (va < vb) {
-                    return -1;
-                }
-                aIndex += 4;
-                bIndex += 4;
+        if (uintCount > 0) {
+            boolean bufferAIsBigEndian = bufferA.order() == ByteOrder.BIG_ENDIAN;
+            final long res;
+            if (bufferA.order() == bufferB.order()) {
+                res = bufferAIsBigEndian ? compareUintBigEndian(bufferA, bufferB, aIndex, bIndex, uintCount) :
+                        compareUintLittleEndian(bufferA, bufferB, aIndex, bIndex, uintCount);
+            } else {
+                res = bufferAIsBigEndian ? compareUintBigEndianA(bufferA, bufferB, aIndex, bIndex, uintCount) :
+                        compareUintBigEndianB(bufferA, bufferB, aIndex, bIndex, uintCount);
             }
-        } else {
-            for (int i = uintCount; i > 0; i --) {
-                long va = bufferA.getUnsignedInt(aIndex);
-                long vb = swapInt(bufferB.getInt(bIndex)) & 0xFFFFFFFFL;
-                if (va > vb) {
-                    return 1;
-                }
-                if (va < vb) {
-                    return -1;
-                }
-                aIndex += 4;
-                bIndex += 4;
+            if (res != 0) {
+                // Ensure we not overflow when cast
+                return (int) Math.min(Integer.MAX_VALUE, res);
             }
+            int uintCountIncrement = uintCount << 2;
+            aIndex += uintCountIncrement;
+            bIndex += uintCountIncrement;
         }
 
-        for (int i = byteCount; i > 0; i --) {
-            short va = bufferA.getUnsignedByte(aIndex);
-            short vb = bufferB.getUnsignedByte(bIndex);
-            if (va > vb) {
-                return 1;
-            }
-            if (va < vb) {
-                return -1;
+        for (int aEnd = aIndex + byteCount; aIndex < aEnd; ++aIndex, ++bIndex) {
+            int comp = bufferA.getUnsignedByte(aIndex) - bufferB.getUnsignedByte(bIndex);
+            if (comp != 0) {
+                return comp;
             }
-            aIndex ++;
-            bIndex ++;
         }
 
         return aLen - bLen;
     }
 
+    private static long compareUintBigEndian(
+            ByteBuf bufferA, ByteBuf bufferB, int aIndex, int bIndex, int uintCount) {
+        for (int aEnd = aIndex + uintCount; aIndex < aEnd; aIndex += 4, bIndex += 4) {
```

Comment:
is `aEnd` correct here? does `uintCount` need to be multiplied by 4 to be in unit of `bytes`?